tools/docker/static-analysis/cppcheck.sh

```diff
@@ -0,0 +1,96 @@
+#!/bin/bash
+
+scriptdir="$(cd "${0%/*}"; pwd)"
+rootdir="${scriptdir%/*/*/*}"
+
+. "${rootdir}/tools/docker/functions.sh"
+
+OUTPUT_FILE="$rootdir/cppcheck_results.txt"
+INCLUDES="\
+-I$rootdir/framework/platform/bpl/include \
+-I$rootdir/framework/transport/ieee1905_transport/include \
+-I$rootdir/framework/common/include \
+-I$rootdir/framework/tlvf/src/include \
+-I$rootdir/framework/tlvf/AutoGenerated/include \
+-I$rootdir/framework/discovery_agent/include \
+-I$rootdir/common/beerocks/btl/include \
+-I$rootdir/common/beerocks/tlvf/src/include \
+-I$rootdir/common/beerocks/tlvf/AutoGenerated/include \
+-I$rootdir/common/beerocks/bcl/include \
+-I$rootdir/common/beerocks/bwl/include"
+
+usage() {
+    echo "usage: $(basename "$0") <source> [source]"
+    echo "run cpp check on <source> and output number of issues per severity"
+}
+
+run_cppcheck() {
+    # --force is needed to check more than 12 ifdefs
+    cppcheck --template='{file}:{line}:{column}: {severity}: {message} [{id}]\n{code}' --force --error-exitcode=1 --enable=warning,style,information -q $INCLUDES "$@" 2>&1 | tee "$OUTPUT_FILE"
+}
+
+colorize_severity() {
+    local severity msg
+    severity="$1"
+    msg="$2"
+    case "$severity" in
+        error)
+            printf '\033[1;31m%s\033[0m\n' "$msg"
+            ;;
+        warning)
+            printf '\033[1;33m%s\033[0m\n' "$msg"
+            ;;
+        style)
+            printf '\033[1;34m%s\033[0m\n' "$msg"
+            ;;
+        performance)
+            printf '\033[1;35m%s\033[0m\n' "$msg"
+            ;;
+        portability)
+            printf '\033[1;36m%s\033[0m\n' "$msg"
+            ;;
+        ok)
+            printf '\033[1;90m%s\033[0m\n' "$msg"
+            ;;
+
+        *)
+            echo unknown severity "$severity"
+            exit 1
+            ;;
+    esac
+}
+
+output_results() {
+    local severities status
+    severities="error warning style performance portability"
+    status=0
+    for s in $severities; do
+        nb_issues="$(grep -Ec "^[^ ]+ $s:" "$OUTPUT_FILE")"
+        if [ $nb_issues -gt 0 ] ; then
+            colorize_severity "$s" "$nb_issues issues with severity $s"
+            case "$s" in
+                # report the following severities in exit code:
+                error|warning)
+                    status=$((status + 1))
+                    ;;
```

Comment:
I think it would be good to also show these on the output, so you don't have to hunt down the artefacts to fix issues. Something as simple as the grep above without the -c should be sufficient.

However, that should probably be done before the summary...

Ideal output would be:
```
WARNING
WARNING
WARNING
ERROR
ERROR
ERROR
3 issues with severity error
3 issues with severity warning
0 issues with severity style
```

i.e., first all warnings, then all errors, then the summary. This makes sure the errors are clearly visible at the end.

(nitpick)

Reply:
You mean you want the issues to be sorted by severity level?
In that case I can wait until cppcheck is done, then output them sorted.
If I keep the current behavior and add the warnings/errors at the end, they will be output 2 times...

Reply:
Never mind, it was absolutely a "nice-to-have". Please merge ASAP.